Feat/task 01 user registration

farchettiensis opened the pull request:
closes issue #1

farchettiensis committed:
chore(dependencies): add dependencies and make migration script (#1)

farchettiensis committed:
feat(user): add UserModel and create users migration (#1)

farchettiensis committed:
chore(lint): update type-only imports in VehicleModel.ts and DealershipModel.ts

farchettiensis committed:
chore(lint): fix unnecessary template literals and unused parameter warnings

farchettiensis committed:
chore(dependency): add Zod dependency (#1)

farchettiensis committed:
feat(users): implement user routes and controllers (#1)

- Added new user routes to router.
- Implemented user controller methods (index, create, store, edit, update) with Zod schemas for validating create and update operations.

farchettiensis committed:
cleanup(usersController): remove unused import

farchettiensis committed:
feat(model): adjust UserModel JSON schema (#1)

- Adjusted required properties in the UserModel JSON schema.

farchettiensis committed:
feat(views): add/update user views (#1)

- Created views for user creation, listing, and updating.
- Updated the shared _form.ejs to conditionally render the password field (only for new users) based on context.

farchettiensis committed:
chore(lint): update config type import

farchettiensis committed:
refactor(issue #1): reorganize project into Clean Architecture layers

- Moved http/controllers/, http/views/, http/router.ts, and http/server.ts to interface/ folder.
- Moved database/database.ts, database/migrations/, and database/models/ to infrastructure/ folder.

farchettiensis committed:
refactor(#1): extract schemas from usersController.ts into a dedicated file

farchettiensis committed:
feat(#1): introduce user DTOs for consistent data transfer

farchettiensis committed:
feat(#1): add user repository

- Created userRepository with create, update, findById, findByEmail, and findAll methods.

farchettiensis committed:
feat(#1): add createUser and updateUser use cases

- Encapsulate the business logic for creating and updating users.
- Use the new userRepository methods and DTOs.
- Return a Result object for consistent error handling.

farchettiensis committed:
refactor(#1): integrate use cases into usersController

- Store (creation) and update methods now delegate to createUser and updateUser use cases.

farchettiensis committed:
chore(#1): update handler import and add result utility

- Adjusted import to satisfy linter's type-only rules.
- Added a Result wrapper to standardize success/failure handling across the codebase.

farchettiensis committed:
chore(#1): add faker, fishery, and vitest dependencies; create vitest.config.ts

- Added test dependencies and vitest configuration file

farchettiensis committed:
feat(#1): update UserModel for optional dealership properties

farchettiensis committed:
feat(#1): add dealership factory

farchettiensis committed:
feat(#1): add user factory

- Created a factory for UserModel with conditional logic.
- Automatically creates a dealership when the user's role is DEALERSHIP and no dealershipId is provided.

farchettiensis committed:
feat(#1): add vehicle factory (WIP)

danielmbrasil commented on src/application/dtos/userDTOs.ts:
```diff
@@ -0,0 +1,22 @@
+  name: string;
+  email: string;
+  password: string;
+  role: 'ADMIN' | 'DEALERSHIP';
```
`role` poderia ser um tipo. Só aqui, você repetiu 3 vezes esse trecho de código.

danielmbrasil commented on src/_lib/http/handler.ts:
```diff
@@ -1,4 +1,4 @@
-import { FastifyReply, FastifyRequest, RouteGenericInterface } from 'fastify';
+import type { FastifyReply, FastifyRequest, RouteGenericInterface } from 'fastify';
```
Não entendi a necessidade dessa alteração. Não vi nenhuma alteração no `tsconfig.json` 🤔. De qualquer forma, em um projeto real, você deveria evitar modificar coisas já existentes no projeto caso não tenha relação/necessidade com a sua tarefa. PS: https://github.com/farchettiensis/w11-node-auth-challenge/commit/f45c98a04f1574ec4526d2a9c369f3dfaa8e5a85#commitcomment-154752350

farchettiensis replied:
Algumas alterações, como essa, foram sugestões do linter - nesse caso, a regra exige que seja feito assim se os imports forem só do tipo.

Eu entendi o seu ponto e você tem razão. Mas você acha que agora teria que voltar ao original?

danielmbrasil replied:
Não precisa, como eu tinha dito, leva como dica para quando trabalhar em um projeto real para evitar esse tipo de alteração desncessária.

danielmbrasil commented on src/application/useCases/user/createUser.ts:
```diff
@@ -0,0 +1,32 @@
+  const existingUser = findResult.data;
+
+  if (existingUser) {
+    return Result.fail({ code: 'INVALID', message: 'Email already in use' });
```
Acredito que os códigos de erros poderiam estar mapeados.

danielmbrasil commented on src/application/useCases/user/updateUser.ts:
```diff
@@ -0,0 +1,42 @@
+
+  const { name, email, role, dealershipId } = input;
+
+  const emailResult = await UserRepository.findByEmail(email);
+
+  if (emailResult.success && emailResult.data) {
+    const existingUser = emailResult.data;
+
+    if (existingUser.id !== id) {
+      return Result.fail({ code: 'INVALID', message: 'Email already in use' });
+    }
+  }
```
Não seria melhor delegar isso para o `update`? Se a constraint falhar, ele deve mostrar algum erro referente a violação da constraint de unicidade e você pode tratar isso de acordo com a regra de negócio, evitando assim de fazer duas queries.

farchettiensis replied:
Simplifiquei os use cases de update e create e atualizei os testes de acordo. Os erros de constraints do banco de dados ficam a cargo do repository apenas.

danielmbrasil commented on src/infrastructure/database/migrations/20250331194230_create_users.ts:
```diff
@@ -0,0 +1,23 @@
+    t.string('email').notNullable().unique();
+    t.string('name').notNullable();
+    t.string('encryptedPassword').notNullable();
+    t.enum('role', ['ADMIN', 'DEALERSHIP']).notNullable();
```
Um problema que vejo de definir regra de negócio no banco, é que se você decidir ter mais um `role` mais tarde, vai precisar fazer uma migração para isso, além da alteração na aplicação.

farchettiensis replied:
Nesse caso, você tem alguma sugestão? Eu vi que no banco ele criou esse campo como "text".

danielmbrasil replied:
O que eu vejo geralmente é uma coluna do tipo string. Na aplicação você define quais strings são válidas para o campos e faz a validação antes disso chegar no banco.
Da maneira que foi feito não é exatemente um problema, chega a ser uma camada a mais de segurança. Mas tem esse ponto que eu mencionei que pode (ou não) ser um problema a levar em consideração.
Pode deixar como está, o meu comentário era mais uma observação mesmo.

danielmbrasil commented on src/infrastructure/database/models/BaseModel.ts:
Entendi que faz sentido a modificação, mas novamente um aviso, em um projeto real você não chega alterando tudo porque discorda de como estão as coisas.

danielmbrasil commented on src/infrastructure/database/models/UserModel.ts:
```diff
@@ -0,0 +1,111 @@
+import { BaseModel } from './BaseModel.js';
+import { DealershipModel, type DealershipSchema } from './DealershipModel.js';
+
+export enum UserRole {
```
Acredito que poderia ter usado isso no DTO.

farchettiensis committed:
chore(config): update database configuration, knexfile, and scripts for dev/test environments (#1)

- Updated src/config.ts to set debug to true for development and false for test.
- Adjusted src/infrastructure/database/database.ts to remove hardcoded debug true, return the knex instance, and add log options for connect/disconnect.
- Refactored knexfile.ts: fixed migrations directory, updated KnexConfig type to Record<string, Knex.Config>, and refactored dbConfig to avoid using spread on accumulators.
- Added new package.json scripts for dropping/creating dev and test databases, running migrations and rollbacks, and running tests (vitest).

danielmbrasil commented on src/infrastructure/database/repositories/userRepository.ts:
```diff
@@ -0,0 +1,64 @@
+  async create(user: UserModel) {
+    try {
+      const createdUser = await UserModel.query().insertAndFetch(user);
+
+      return Result.succeed(createdUser);
+    } catch (error) {
+      const message = error instanceof Error ? error.message : 'Database error';
+
+      return Result.fail<UserModel>({ code: 'DATABASE_ERROR', message });
+    }
```
Acho que o tratamento aqui poderia ser um pouco melhor e identificar o tipo de erro, por exemplo, violação de constraint de unicidade.

farchettiensis replied:
Nesse caso eu teria que mapear os erros do postgres?

danielmbrasil replied:
Não necessariamente. Vi aqui que o Objection tem alguns erros específicos como `ValidationError`, `NotFoundError` e `UniqueValidationError` (https://vincit.github.io/objection.js/api/objection/#uniqueviolationerror). Você poderia pegar o tipo de erro e fazer um tratamento mais adequeado. Agora, não sei te dizer como fazer uma mensagem de erro tipo "email já está em uso" porque a documentação não está muito clara sobre a estrutura do erro retornado. Você poderia dar uma olhada nisso e ver se tem como fazer um tratamento mais específico. Acredito que ele deve retornar o código do erro que é definido pelo Postgres (https://www.postgresql.org/docs/current/errcodes-appendix.html) e algum detalhe sobre qual campo jogou o erro. Mas, para simplificar sua vida, poderia considerar que um erro de `UniqueValidationError` vai ser sempre no email já que a única constraint para essa tabela.

danielmbrasil commented on src/infrastructure/database/repositories/userRepository.ts:
```diff
@@ -0,0 +1,64 @@
+  },
+
+  async update(id: number, user: UserModel) {
+    try {
+      const updatedUser = await UserModel.query().patchAndFetchById(id, user);
+
+      return Result.succeed(updatedUser);
+    } catch (error) {
+      const message = error instanceof Error ? error.message : 'Database error';
+
+      return Result.fail<UserModel>({ code: 'DATABASE_ERROR', message });
+    }
```
Mesma coisa aqui.

danielmbrasil commented on src/interface/http/controllers/dealershipController.ts:
```diff
@@ -17,15 +19,19 @@ const store = handler<{ Body: { name: string } }>(async (request, reply) => {
     await DealershipModel.query().insert({ name });
 
-    return reply.redirect(`/dealerships`);
+    return reply.redirect('dealerships');
```
Isso daqui funciona? Você removeu a barra `/` também.

farchettiensis committed:
feat(dto): update UpdateUserDTO to accept null for dealershipId and remove unused interface (#1)

- Adjusted UpdateUserDTO to allow dealershipId to be number or null,
  reflecting that admin users should have null dealership fields.

danielmbrasil commented on src/interface/http/controllers/usersController.ts:
```diff
@@ -0,0 +1,94 @@
+} from '../../schemas/userSchemas.js';
+
+const index = handler(async (_request, reply) => {
+  const users = await UserModel.query();
```
Aqui poderia usar o `findAll` do repositório, ou criar uma action para isso.

farchettiensis committed:
feat(repository): update userRepository methods to eagerly load dealership (#1)

farchettiensis committed:
feat(factory): adjust VehicleFactory for proper dealership association (#1)

farchettiensis committed:
test: add test setups and use case tests for createUser and updateUser (#1)

- Added setupIntegrationTest.ts to manage database connection, cleanup, and teardown.
- Created tests for the createUser use case covering both dealership and admin scenarios.
- Created tests for the updateUser use case, including tests for duplicate email handling.

danielmbrasil commented on src/interface/http/controllers/vehiclesController.ts:
Nesse controller parece que você esqueceu de seguir o padrão de repoistório e action.

farchettiensis replied:
Na verdade, eu deixei como estava. Eu vi que ia gastar bastante tempo refatorando os controllers de dealership e vehicles, então deixei _on hold_ por ora. Na verdade, acho que deveria ter feito isso com user só no final :(

danielmbrasil commented on src/interface/schemas/userSchemas.ts:
```diff
@@ -0,0 +1,51 @@
+    name: z.string().nonempty(),
+    email: z.string().email(),
+    password: z.string().nonempty(),
+    role: z.enum(['ADMIN', 'DEALERSHIP']),
```
Poderia evitar a duplicação usando o enum do model.

danielmbrasil commented on tests/application/useCases/user/createUser.test.ts:
```diff
@@ -0,0 +1,114 @@
+
+  describe('when the data is valid', () => {
+    describe('and the user is a dealership', () => {
+      it('creates a new dealership user', async () => {
```
Acho que falta o caso de erro onde `delearshipId` não é passado.

farchettiensis replied:
Adicionado, tive que fazer um try-catch no use case, porque parece que se não o error bubble up pra algum lugar e gera um retorno diferente do Result.fail.

danielmbrasil commented on tests/application/useCases/user/createUser.test.ts:
```diff
@@ -0,0 +1,114 @@
+            id: expect.any(Number),
+            name: userData.name,
+            email: userData.email,
+            encryptedPassword: expect.any(String),
```
Isso daqui seria uma falha de segurança. Mesmo que a senha tenha sido criptografada, você não deveria retornar o hash.

farchettiensis replied:
Acho que porque nesse projeto não tem um serializer ou um formato de resposta definido. Acho que seria essa a forma de evitar essa falha, não?

danielmbrasil replied:
Sim, um serializer poderia ser uma opção mas você também pode usar o `$formatJson` para omitir esses campos diretamente no model: https://github.com/Vincit/objection.js/issues/1851#issuecomment-698813304.

EDIT: não sei se fazer o override desse método vai ter consequências em outros selects. Talvez, nesse caso, o melhor seria um serializer mesmo.

danielmbrasil commented on tests/application/useCases/user/updateUser.test.ts:
```diff
@@ -0,0 +1,117 @@
+    await integrationTest.tearDown();
+  });
+
+  describe('when the data is valid', () => {
```
Falta o caso de teste onde o usuário não é encontrado e também quando algum dado é inválido.

farchettiensis replied:
Adicionei um cenário e as demais validações ficam a cargo do repository e do controller.

farchettiensis committed:
feat(#1): add withGraphFetched to UserRepository methods

- Updated findById and findByEmail in src/infrastructure/database/repositories/userRepository.ts
  to eagerly load the dealership relation via .withGraphFetched('dealership').

farchettiensis committed:
test(#1): add tests for UserRepository

- Added tests covering all repository methods

farchettiensis committed:
refactor(#1): create UserRoleType type to be used in the DTOs, avoiding repetition

farchettiensis committed:
feat(#1): standardize error handling with ApplicationError and DatabaseError

- Introduced ApplicationError and DatabaseError classes in src/errors/applicationError.ts and src/errors/databaseError.ts.
- Created errorCodes.ts with a centralized ErrorCodes enum.
- Updated createUser.ts and updateUser.ts to return failures using ApplicationError/DatabaseError with mapped error codes.
- Refactored UserRepository and its tests (userRepository.ts, userRepository.test.ts) to use the new error types.
- Adjusted createUser.test.ts and updateUser.test.ts to expect the standardized error structure.

farchettiensis committed:
fix(#1): fix import error in createUser.ts and updateUser.ts

farchettiensis committed:
refactor(#1): refactor Zod user schemas to use enum directly instead repeating the values

farchettiensis committed:
feat(#1): add test for missing dealershipId and catch validation errors

- Added a test scenario to verify that creating a dealership user without a dealershipId fails.
- Wrapped UserModel.fromJson() in a try/catch block in createUser use case to catch and map validation errors.

farchettiensis committed:
feat(#1): add test for missing user

farchettiensis committed:
fix(#1): fix redirect route for dealershipController.store

farchettiensis committed:
feat(#1): refactor user queries to utilize use cases and update repository error handling

- Updated UsersController index and edit methods to delegate to findUsers/findUser use cases instead of direct queries.
- Created findUsers and findUser use cases to encapsulate user retrieval logic.
- Modified UserRepository.findById to return Result.fail with NOT_FOUND when a user isn't found.
- Updated userRepository tests to reflect the new behavior.

farchettiensis committed:
feat(#1): streamline user creation and update use cases and repository error handling

- Refactored createUser and updateUser use cases to remove duplicate validations, relying on repository/database constraints and errors caught by the repository.
- Updated createUser and updateUser test suites to expect errors from the repository layer.
- Modified UserRepository.findByEmail to return a Result.fail with a NOT_FOUND error when a user is not found.
- Aligned error messages in findById and findByEmail for consistency.
- Adjusted userRepository.test.ts descriptions to document new expectations and the assertions.

farchettiensis committed:
test(#1): add tests for UsersController (WIP)

- Added tests for index, create, edit, store, and update actions.
- Covered success and error scenarios, including validation failures and error mapping.
- Ensured controllers correctly delegate to use cases and handle repository errors.

farchettiensis committed:
test(#1): fix update test

- The test was failing because the dealershipId is transformed into a number before being saved
